aws/resource_aws_codedeploy_deployment_config.go

```diff
@@ -137,15 +226,102 @@ func expandAwsCodeDeployConfigMinimumHealthHosts(d *schema.ResourceData) *codede
 	return &minimumHealthyHost
 }
 
+func expandAwsCodeDeployTrafficRoutingConfig(d *schema.ResourceData) *codedeploy.TrafficRoutingConfig {
+	block, ok := d.GetOk("traffic_routing_config")
+	if !ok {
+		return nil
+	}
+	config := block.([]interface{})[0].(map[string]interface{})
+	trafficRoutingConfig := codedeploy.TrafficRoutingConfig{}
+
+	if trafficType, ok := config["type"]; ok {
+		trafficRoutingConfig.SetType(trafficType.(string))
+	}
+	if canary, ok := config["time_based_canary"]; ok && len(canary.([]interface{})) > 0 {
+		canaryConfig := canary.([]interface{})[0].(map[string]interface{})
+		trafficRoutingConfig.SetTimeBasedCanary(expandAwsCodeDeployTrafficTimeBasedCanaryConfig(canaryConfig))
+	}
+	if linear, ok := config["time_based_linear"]; ok && len(linear.([]interface{})) > 0 {
+		linearConfig := linear.([]interface{})[0].(map[string]interface{})
+		trafficRoutingConfig.SetTimeBasedLinear(expandAwsCodeDeployTrafficTimeBasedLinearConfig(linearConfig))
+	}
+
+	return &trafficRoutingConfig
+}
+
+func expandAwsCodeDeployTrafficTimeBasedCanaryConfig(config map[string]interface{}) *codedeploy.TimeBasedCanary {
+	canary := codedeploy.TimeBasedCanary{}
+	if interval, ok := config["interval"]; ok {
+		canary.SetCanaryInterval(int64(interval.(int)))
+	}
+	if percentage, ok := config["percentage"]; ok {
+		canary.SetCanaryPercentage(int64(percentage.(int)))
+	}
+	return &canary
+}
+
+func expandAwsCodeDeployTrafficTimeBasedLinearConfig(config map[string]interface{}) *codedeploy.TimeBasedLinear {
+	linear := codedeploy.TimeBasedLinear{}
+	if interval, ok := config["interval"]; ok {
+		linear.SetLinearInterval(int64(interval.(int)))
+	}
+	if percentage, ok := config["percentage"]; ok {
+		linear.SetLinearPercentage(int64(percentage.(int)))
+	}
+	return &linear
+}
+
 func flattenAwsCodeDeployConfigMinimumHealthHosts(hosts *codedeploy.MinimumHealthyHosts) []map[string]interface{} {
 	result := make([]map[string]interface{}, 0)
+	if hosts == nil {
+		return result
+	}
 
 	item := make(map[string]interface{})
 
 	item["type"] = *hosts.Type
 	item["value"] = *hosts.Value
 
-	result = append(result, item)
+	return append(result, item)
+}
+
+func flattenAwsCodeDeployTrafficRoutingConfig(config *codedeploy.TrafficRoutingConfig) []map[string]interface{} {
+	result := make([]map[string]interface{}, 0)
+	if config == nil {
+		return result
+	}
+
+	item := make(map[string]interface{})
+
+	item["type"] = *config.Type
```

Comment:
To prevent potential panics, we prefer to use the AWS Go SDK provided methods for dereferencing API responses, e.g.

```suggestion
	item["type"] = aws.StringValue(config.Type)
```

Same applies to similar `*` dereferences below.